File: Ch. 1/IsUnique.java
import java.util.*;

/**
 * Check if a string has unique characters
 */

public class IsUnique {

   public static void main (String[] args) {
      Scanner scanner = new Scanner(System.in);
      String s = scanner.next();
      System.out.println(unique_1(s));
      System.out.println(unique_2(s));
   }

   /* Hashmap O(N) */
   public static boolean unique_1(String s) {
      HashMap<Character, Integer> map = new HashMap<>();
      char[] c = s.toCharArray();
      for (int i = 0; i < c.length; i++) {
         if (map.containsKey(c[i])) {
            return false;
         }
         else {
            map.put(c[i], 1);
         }
      }
      return true;
   }

   /* Sort O(N log N) */
   public static boolean unique_2(String s) {
      char[] c = s.toCharArray();
      Arrays.sort(c);
      for (int i = 0; i < c.length-1; i++) {
         if (c[i] == c[i+1]) {
            return false;
         }
      }
      return true;
   }
}

File: Ch. 1/IsUnique.py
# Check if a string has unique characters

# dictionary O(n)
def unique_1(check_str):
    freq_dict = {}
    
    for char in check_str:
        if char in freq_dict:
            return False
        else:
            freq_dict[char] = 1
    return True

# sort O(n log n)
def unique_2(check_str):
    sorted_str_list = sorted(check_str)

    for i in range(len(sorted_str_list)-1):
        if sorted_str_list[i] == sorted_str_list[i+1]:
            return False
    return True

test_str = input('Enter string: ')
print(unique_1(test_str))
print(unique_2(test_str))

